examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,2035 @@
+        disable=not accelerator.is_local_main_process,
+    )
+
+    for epoch in range(first_epoch, args.num_train_epochs):
+        for step, batch in enumerate(train_dataloader):
+            with accelerator.accumulate(unet):
+                # 1. Load and process the image and text conditioning
+                if args.use_image_conditioning:
+                    image, text, cond_image = batch
+                else:
+                    image, text = batch
+
+                image = image.to(accelerator.device, non_blocking=True)
+                encoded_text = compute_embeddings_fn(text)
+                if args.use_image_conditioning:
+                    encoded_image = compute_image_embeddings_fn(cond_image)
+
+                pixel_values = image.to(dtype=weight_dtype)
+                if vae.dtype != weight_dtype:
+                    vae.to(dtype=weight_dtype)
+
+                # encode pixel values with batch size of at most args.vae_encode_batch_size
+                latents = []
+                for i in range(0, pixel_values.shape[0], args.vae_encode_batch_size):
+                    latents.append(vae.encode(pixel_values[i : i + args.vae_encode_batch_size]).latent_dist.sample())
+                latents = torch.cat(latents, dim=0)
+
+                latents = latents * vae.config.scaling_factor
+                latents = latents.to(weight_dtype)
+                bsz = latents.shape[0]
+
+                # 2. Sample random student timesteps s uniformly in `student_timestep_schedule` and sample random
+                # teacher timesteps t uniformly in [0, ..., noise_scheduler.config.num_train_timesteps - 1].
+                student_index = torch.randint(0, student_distillation_steps, (bsz,), device=latents.device).long()
+                student_timesteps = student_timestep_schedule[student_index]
+                teacher_timesteps = torch.randint(
+                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=latents.device
+                ).long()
+
+                # 3. Sample noise and add it to the latents according to the noise magnitude at each student timestep
+                # (that is, run the forward process on the student model)
+                student_noise = torch.randn_like(latents)
+                noisy_student_input = noise_scheduler.add_noise(latents, student_noise, student_timesteps)
+
+                # 4. Prepare prompt embeds (for teacher/student U-Net) and text embedding (for discriminator).
+                prompt_embeds = encoded_text.pop("prompt_embeds")
+                text_embedding = encoded_text.pop("text_embedding")
+                image_embedding = None
+                if args.use_image_conditioning:
+                    image_embedding = encoded_image.pop("image_embeds").float()
+                    # Only supply image conditioning when student timestep is not last training timestep T.
+                    image_embedding = torch.where(
```

Comment:
currently the same masked `image_embedding` is fed to discriminator for both real and fake images, but it seems for which real image its `image_embedding` being masked out could be random with a rate say `1 / num_inference_steps` instead of depending on `student_timesteps`, not sure if this could make big difference though

Reply:
Yeah same. I am sure either how impactful this would be.